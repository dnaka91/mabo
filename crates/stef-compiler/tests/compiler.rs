use std::{
    fmt::{self, Display},
    fs,
};

use insta::{assert_snapshot,with_settings, glob};
use miette::{Diagnostic, MietteHandler, MietteHandlerOpts, NamedSource, Report, ReportHandler};
use stef_parser::Schema;

#[test]
fn compile_invalid_schema() {
    struct Wrapper<'a>(&'a MietteHandler, &'a dyn Diagnostic);

    impl<'a> Display for Wrapper<'a> {
        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
            self.0.debug(self.1, f)
        }
    }

    let handler = MietteHandlerOpts::new()
        .color(false)
        .context_lines(3)
        .force_graphical(true)
        .terminal_links(false)
        .width(120)
        .build();

    glob!("inputs/invalid/*.stef", |path| {
        let name = path.file_stem().unwrap().to_str().unwrap();
        let input = fs::read_to_string(path).unwrap();
        let schema = Schema::parse(input.as_str()).unwrap();
        let result = stef_compiler::validate_schema(name, &schema).unwrap_err();
        let report = Report::new(result).with_source_code(NamedSource::new(
            path.file_name().unwrap().to_string_lossy(),
            input.clone(),
        ));

        with_settings!({
            description => input.trim(),
            omit_expression => true,
        }, {
            assert_snapshot!("error", Wrapper(&handler, &*report).to_string());
        });
    });
}
