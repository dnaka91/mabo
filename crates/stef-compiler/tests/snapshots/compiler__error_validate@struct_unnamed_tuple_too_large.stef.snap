---
source: crates/stef-compiler/tests/compiler.rs
description: "struct Sample((u8, u8, u8, u8, u8, u8, u8, u8, u8, u8, u8, u8, u8) @1)"
input_file: crates/stef-compiler/tests/inputs/validate/struct_unnamed_tuple_too_large.stef
---
  × invalid tuple element size found
  ╰─▶ tuples with ❬W❭more than 12 elements❬W❭ are invalid
   ╭─[struct_unnamed_tuple_too_large.stef:1:1]
 1 │ struct Sample((u8, u8, u8, u8, u8, u8, u8, u8, u8, u8, u8, u8, u8) @1)
   ·               ──────────────────────────┬─────────────────────────
   ·                                         ╰── declared here
   ╰────
  help: a tuple must have between 2 and 12 elements

